File: lib/llm-service.ts
/**
 * LLM prompt construction and README generation.
 *
 * Builds a dense, structured prompt from the RepoAnalysis data and sends
 * it to OpenAI (or compatible) to produce a comprehensive README.
 */

import type { RepoAnalysis } from "./github-service"

// ---------------------------------------------------------------------------
// Prompt helpers
// ---------------------------------------------------------------------------

function formatTree(tree: string[], maxLines = 120): string {
  if (tree.length <= maxLines) return tree.join("\n")
  const half = Math.floor(maxLines / 2)
  const omitted = tree.length - maxLines
  return (
    tree.slice(0, half).join("\n") +
    `\n\n... (${omitted} more files omitted) ...\n\n` +
    tree.slice(-half).join("\n")
  )
}

function formatKeyFiles(keyFiles: Record<string, string>): string {
  const entries = Object.entries(keyFiles)
  if (entries.length === 0)
    return "(No key configuration files found at the repository root.)"

  return entries
    .map(([name, content]) => {
      const truncated =
        content.length > 6000
          ? content.slice(0, 6000) + "\n\n[...truncated...]"
          : content
      return `### ${name}\n\`\`\`\n${truncated}\n\`\`\``
    })
    .join("\n\n")
}

// ---------------------------------------------------------------------------
// System prompt
// ---------------------------------------------------------------------------

const SYSTEM_PROMPT = `You are an expert technical writer specializing in open-source documentation.
Given structured data about a GitHub repository (metadata, file tree, key file
contents, and license information), generate a **comprehensive, well-organized
README.md** in valid GitHub-Flavored Markdown.

The README **must** contain ALL of the following sections in this order:

1. **Title & Badges**
   - Project name as an H1.
   - A short tagline / one-liner description.
   - Shields.io badges for: build status (if CI config is found), license,
     language/version, and stars.

2. **Introduction**
   - A 2-3 paragraph overview of what the project does, who it is for,
     and what problems it solves.

3. **Why This Project? (Theory / Motivation)**
   - Explain the "why" behind the project — the gap it fills, the theory
     or principles it builds on, and how it differs from alternatives.

4. **Features**
   - A bullet list of the project's key capabilities.

5. **Project Structure**
   - An ASCII-style directory tree showing the most important files and
     folders (max ~25 entries). Annotate each entry briefly.

6. **Getting Started / How to Run Locally**
   - Prerequisites (runtime version, tools).
   - Step-by-step installation commands.
   - How to start the development server or run the main entry point.
   - Example usage or first command to try.

7. **Configuration** (if applicable)
   - Describe any configuration files, environment variables, or flags.

8. **API Reference / Usage** (if applicable)
   - Summarize key exports, CLI commands, or API endpoints.

9. **Contributing**
   - Brief contributing guidelines (fork -> branch -> PR workflow).

10. **Authors**
    - Credit the repository owner. If contributor info is not available,
      use the GitHub owner as the primary author.

11. **License**
    - State the license name and link to the LICENSE file.

Rules:
- Output ONLY raw Markdown. No commentary, no "\`\`\`markdown" wrapper.
- Use proper GFM: fenced code blocks with language hints, tables where
  appropriate, task lists if relevant.
- Infer technologies, frameworks, and patterns from the file tree and
  key file contents. Do not hallucinate features that are not evidenced.
- Keep the tone professional but approachable.
- Prefer concise sentences. Avoid filler.`

// ---------------------------------------------------------------------------
// Build the user message
// ---------------------------------------------------------------------------

export function buildPrompt(analysis: RepoAnalysis): string {
  const { meta } = analysis
  const topicsStr = meta.topics.length > 0 ? meta.topics.join(", ") : "none"

  let licenseSection: string
  if (analysis.licenseName) {
    licenseSection = `Detected license: ${analysis.licenseName}`
  } else if (analysis.licenseContent) {
    licenseSection =
      "A LICENSE file was found but its SPDX identifier is unknown. " +
      "Here are the first 500 characters:\n" +
      analysis.licenseContent.slice(0, 500)
  } else {
    licenseSection = "No license file detected."
  }

  return `Analyze the following GitHub repository and generate a comprehensive,
production-quality README.md in Markdown.

${"=".repeat(60)}
REPOSITORY METADATA
${"=".repeat(60)}
- Name: ${meta.fullName}
- Description: ${meta.description ?? "(none)"}
- Primary language: ${meta.language ?? "unknown"}
- Stars: ${meta.stars.toLocaleString()}  |  Forks: ${meta.forks.toLocaleString()}
- Topics: ${topicsStr}
- Homepage: ${meta.homepage ?? "(none)"}
- Default branch: ${meta.defaultBranch}

${"=".repeat(60)}
FILE TREE (top ${analysis.tree.length} files)
${"=".repeat(60)}
${formatTree(analysis.tree)}

${"=".repeat(60)}
KEY FILES (contents)
${"=".repeat(60)}
${formatKeyFiles(analysis.keyFiles)}

${"=".repeat(60)}
LICENSE
${"=".repeat(60)}
${licenseSection}`
}

// ---------------------------------------------------------------------------
// Generation (streaming)
// ---------------------------------------------------------------------------

export async function* generateReadmeStream(
  analysis: RepoAnalysis,
  apiKey: string,
  model = "gpt-4o",
): AsyncGenerator<string> {
  const userPrompt = buildPrompt(analysis)

  const resp = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      messages: [
        { role: "system", content: SYSTEM_PROMPT },
        { role: "user", content: userPrompt },
      ],
      temperature: 0.4,
      max_tokens: 4096,
      stream: true,
    }),
  })

  if (!resp.ok) {
    const error = await resp.text()
    throw new Error(`OpenAI API error (${resp.status}): ${error}`)
  }

  const reader = resp.body?.getReader()
  if (!reader) throw new Error("No response body from OpenAI")

  const decoder = new TextDecoder()
  let buffer = ""

  while (true) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split("\n")
    buffer = lines.pop() ?? ""

    for (const line of lines) {
      const trimmed = line.trim()
      if (!trimmed.startsWith("data: ")) continue
      const data = trimmed.slice(6)
      if (data === "[DONE]") return
      try {
        const parsed = JSON.parse(data)
        const content = parsed.choices?.[0]?.delta?.content
        if (content) yield content
      } catch {
        // Skip malformed JSON chunks
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Generation (non-streaming)
// ---------------------------------------------------------------------------

export async function generateReadme(
  analysis: RepoAnalysis,
  apiKey: string,
  model = "gpt-4o",
): Promise<string> {
  const userPrompt = buildPrompt(analysis)

  const resp = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      messages: [
        { role: "system", content: SYSTEM_PROMPT },
        { role: "user", content: userPrompt },
      ],
      temperature: 0.4,
      max_tokens: 4096,
    }),
  })

  if (!resp.ok) {
    const error = await resp.text()
    throw new Error(`OpenAI API error (${resp.status}): ${error}`)
  }

  const data = await resp.json()
  return data.choices?.[0]?.message?.content ?? ""
}

File: backend/services/llm_service.py
"""Gemini-powered README generation service."""

from __future__ import annotations

import logging
from typing import AsyncIterator

import google.generativeai as genai
from google.api_core import exceptions as google_exceptions

from .github_service import RepoAnalysis

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

GEMINI_MODEL = "gemini-1.5-flash"


def _build_prompt(analysis: RepoAnalysis) -> str:
    """Build a dense prompt from repository analysis data."""
    meta = analysis.meta
    key_files = list(analysis.key_files.keys())
    tree_preview = "\n".join(f"  - {path}" for path in analysis.tree[:40])
    if len(analysis.tree) > 40:
        tree_preview += f"\n  - ... ({len(analysis.tree) - 40} more files)"

    key_file_snippets = ""
    for name, content in list(analysis.key_files.items())[:6]:
        snippet = content[:2000]
        key_file_snippets += f"\n### {name}\n```\n{snippet}\n```\n"

    return f"""You are an expert technical writer. Generate a comprehensive, production-ready README.md for this GitHub repository.

Repository: {meta.full_name}
Owner: {meta.owner}
Description: {meta.description or "No description provided"}
Primary language: {meta.language or "Unknown"}
Stars: {meta.stars}
Forks: {meta.forks}
Topics: {", ".join(meta.topics) if meta.topics else "None"}
License: {analysis.license_name or "Not detected"}
Default branch: {meta.default_branch}

File tree (sample):
{tree_preview}

Key configuration files found: {", ".join(key_files) if key_files else "None"}

Key file contents:
{key_file_snippets or "No key files available."}

Requirements:
- Output ONLY valid Markdown (no code fences wrapping the entire README).
- Include: title, badge-style metadata line, introduction, features, tech stack, project structure (ASCII tree), prerequisites, installation/setup, usage, contributing (brief), authors, license.
- Infer setup commands from detected files (package.json → npm, requirements.txt → pip, etc.).
- Use the actual repo name, owner, and detected license.
- Be specific to this repository — avoid generic filler.
- Use clear headings and emoji sparingly for section headers.
- Keep it professional and developer-friendly."""


def _extract_text(response) -> str:
    """Extract text from a Gemini response, raising on blocked/empty output."""
    if not response.candidates:
        raise ValueError("Gemini returned no candidates. The response may have been blocked.")

    candidate = response.candidates[0]
    if not candidate.content or not candidate.content.parts:
        raise ValueError("Gemini returned an empty response.")

    return "".join(part.text for part in candidate.content.parts if part.text)


def _handle_gemini_error(exc: Exception) -> None:
    """Translate Gemini API errors into user-friendly messages."""
    if isinstance(exc, google_exceptions.InvalidArgument):
        raise ValueError("Invalid Gemini API key. Please check your key and try again.") from exc
    if isinstance(exc, google_exceptions.PermissionDenied):
        raise ValueError("Gemini API key denied. Ensure your key is valid and has API access enabled.") from exc
    if isinstance(exc, google_exceptions.ResourceExhausted):
        raise ValueError("Gemini API rate limit exceeded. Please wait and try again.") from exc
    if isinstance(exc, google_exceptions.Unauthenticated):
        raise ValueError("Gemini API authentication failed. Please verify your API key.") from exc
    raise exc


def generate_readme(analysis: RepoAnalysis, api_key: str) -> str:
    """Generate a full README.md using the Gemini API."""
    genai.configure(api_key=api_key)
    model = genai.GenerativeModel(GEMINI_MODEL)
    prompt = _build_prompt(analysis)

    try:
        response = model.generate_content(prompt)
        return _extract_text(response).strip()
    except Exception as exc:
        logger.error("Gemini generation failed: %s", exc)
        _handle_gemini_error(exc)
        raise


async def generate_readme_stream(analysis: RepoAnalysis, api_key: str) -> AsyncIterator[str]:
    """Stream README generation chunks from Gemini."""
    genai.configure(api_key=api_key)
    model = genai.GenerativeModel(GEMINI_MODEL)
    prompt = _build_prompt(analysis)

    try:
        response = model.generate_content(prompt, stream=True)
        for chunk in response:
            if chunk.text:
                yield chunk.text
    except Exception as exc:
        logger.error("Gemini streaming failed: %s", exc)
        _handle_gemini_error(exc)
        raise
